File: reference/game/src/game/entities/base/combat/HitResult.ts
import type { HitboxConfig, StaggerLevel } from "../../../types/CombatTypes"
import type { CombatStats } from "./CombatStats"

export type HitResult = {
  damage: number
  poiseDamage: number
  staggerDuration: number
  staggerLevel: StaggerLevel
  knockbackX: number
  knockbackY: number
  brokePoise: boolean
  killed: boolean
}

export function calculateHit(
  attackerStats: CombatStats,
  defenderStats: CombatStats,
  hitConfig: HitboxConfig,
  damageMultiplier: number,
): HitResult {
  const baseDamage = Math.max(0, attackerStats.attack - defenderStats.defense * 0.5)
  const rawDamage = baseDamage + hitConfig.damage
  const finalDamage = Math.round(rawDamage * damageMultiplier)

  const defenderPoise = defenderStats.poise
  const poiseDamage = hitConfig.poiseDamage
  const brokePoise = poiseDamage >= defenderPoise

  return {
    damage: finalDamage,
    poiseDamage,
    staggerDuration: hitConfig.staggerDuration,
    staggerLevel: hitConfig.staggerLevel,
    knockbackX: hitConfig.knockbackX ?? 0,
    knockbackY: hitConfig.knockbackY ?? 0,
    brokePoise,
    killed: defenderStats.health - finalDamage <= 0,
  }
}

File: internal/combat/hit_result.go
package combat

import "math"

type StaggerLevel int

const (
	StaggerNone StaggerLevel = iota
	StaggerFlinch
	StaggerStagger
	StaggerKnockdown
	StaggerLaunch
)

type HitResult struct {
	Damage          float64
	PoiseDamage     float64
	StaggerDuration float64
	StaggerLevel    StaggerLevel
	KnockbackX      float64
	KnockbackY      float64
	BrokePoise      bool
	Killed          bool
	HitLocation     HurtboxType
}

type HitConfig struct {
	Damage       float64
	PoiseDamage  float64
	StaggerLevel StaggerLevel
	KnockbackX   float64
	KnockbackY   float64
}

func CalculateHit(attack, defense float64, config HitConfig, dmgMult, strength float64) HitResult {
	raw := math.Round((attack*strength - defense*0.5) + config.Damage)
	damage := raw * dmgMult
	if damage < 0 {
		damage = 0
	}
	return HitResult{
		Damage:          damage,
		PoiseDamage:     config.PoiseDamage,
		StaggerDuration: staggerDurationFromLevel(config.StaggerLevel),
		StaggerLevel:    config.StaggerLevel,
		KnockbackX:      config.KnockbackX,
		KnockbackY:      config.KnockbackY,
	}
}

func staggerDurationFromLevel(level StaggerLevel) float64 {
	switch level {
	case StaggerNone:
		return 0
	case StaggerFlinch:
		return 0.2
	case StaggerStagger:
		return 0.5
	case StaggerKnockdown:
		return 1.0
	case StaggerLaunch:
		return 1.5
	default:
		return 0
	}
}
